Refactor certain doctests towards either reStructuredText documentation or Python tests (#464)
crate/crate-python: docs/by-example/sqlalchemy/inspection-reflection.rst

=====================================================

> SQLAlchemy: Database schema inspection and reflection

Reviewer: Is anything of this section CrateDB specific? Looks like it's mostly explaining SQLAlchemy, in which case we could refer to the upstream docs instead and test the functionality in python tests

Author: Hi. While the SQLAlchemy API methods per se are not specific to CrateDB, I found this section worth to keep, to display different kinds of inspection and reflection methods of SQLAlchemy together with their slightly CrateDB-specific outcomes, also because they are not mentioned elsewhere in our documentation (modulo `has_table`), and also because the corresponding spots are a bit obfuscated within the SQLAlchemy documentation.

In other words, while bringing those snippets together from different parts of our documentation, I have made an extra effort to discriminate between the different kinds of introspection/reflection features offered by SQLAlchemy:

- SQLAlchemy inspector
- Schema-supported reflection
- CrateDialect-specific reflection methods

The result is a reasonably concise overview about them, paired with CrateDB-specific responses.

```python
>>> inspector.get_schema_names()
['blob', 'doc', 'information_schema', 'pg_catalog', 'sys']

>>> inspector.default_schema_name
'doc'

>>> dialect.has_schema(connection, 'doc')
True
```

=====================================================

> This section of the documentation, related to CrateDB's SQLAlchemy integration,
> focuses on database schema inspection and reflection features. That is, given
> that you connect to an existing database, how to introspect its metadata
> information to retrieve table- and view-names, and table column metadata.

Reviewer: ```suggestion
This section shows you how to inspect the schema of a database using CrateDB's
SQLAlchemy integration.
```

https://developers.google.com/style/tone

Introduction
============

> The `runtime inspection API`_ provides the ``inspect()`` function, which
> delivers runtime information about a wide variety of SQLAlchemy objects, both
> within SQLAlchemy Core and the SQLAlchemy ORM.
> 
> The ``CrateDialect`` instance provides metadata about the CrateDB cluster,
> like version and schema information.
> 
> As always, let's start with creating an ``Engine`` instance.
> 
>     >>> import sqlalchemy as sa
>     >>> engine = sa.create_engine(f"crate://{crate_host}")
> 
> 
> Inspector
> =========
> 
> The `SQLAlchemy inspector`_ is a low level interface which provides a
> backend-agnostic system of loading lists of schema, table, column, and
> constraint descriptions from a given database.

Reviewer: ```suggestion

The CrateDB SQLAlchemy integration provides different ways to inspect the database.

1) The `runtime inspection API`_ allows you to get an ``Inspector`` instance that can be used to fetch schema names, table names and other information.

2) Reflection capabilities allow you to create ``Table`` instances from existing tables to inspect their columns and constraints.

3) A ``CrateDialect`` allows you to get connection information and it contains low level function to check the existence of schemas and tables.

All approaches require an ``Engine`` instance, which you can create like this:

    >>> import sqlalchemy as sa
    >>> engine = sa.create_engine(f"crate://{crate_host}")

Inspector
=========

The `SQLAlchemy inspector`_ is a low level interface which provides a
backend-agnostic system of loading lists of schema, table, column, and
constraint descriptions from a given database. You create an inspector like
this:
```

    >>> inspector = sa.inspect(engine)

List all schemas:

    >>> inspector.get_schema_names()
    ['blob', 'doc', 'information_schema', 'pg_catalog', 'sys']

List all tables:

    >>> set(['characters', 'cities', 'locations']).issubset(inspector.get_table_names())
    True

    >>> set(['checks', 'cluster', 'jobs', 'jobs_log']).issubset(inspector.get_table_names(schema='sys'))
    True

List all views:

    >>> inspector.get_view_names()
    ['characters_view']

Get default schema name:

    >>> inspector.default_schema_name
    'doc'

Schema-supported reflection
===========================

> A ``Table`` object can be instructed to load information about itself from the
> corresponding database schema object already existing within the database. This
> process is called *reflection*, see `reflecting database objects`_.

Reviewer: ```suggestion
A ``Table`` object can load its own schema information from the corresponding
table in the database. This process is called *reflection*, see `reflecting
database objects`_.
```

In the most simple case you need only specify the table name, a ``MetaData``
object, and the ``autoload_with`` argument.

Create a SQLAlchemy table object:

    >>> meta = sa.MetaData()
    >>> table = sa.Table(
    ...     "characters", meta,
    ...     autoload=True,
    ...     autoload_with=engine)

Reflect column data types from the table metadata:

    >>> table.columns.get('name')
    Column('name', String(), table=<characters>)

    >>> table.primary_key
    PrimaryKeyConstraint(Column('id', String(), table=<characters>, primary_key=True...

CrateDialect
============

After initializing the dialect instance with a connection instance,

    >>> from crate.client.sqlalchemy.dialect import CrateDialect
    >>> dialect = CrateDialect()

    >>> connection = engine.connect()
    >>> dialect.initialize(connection)

the database server version and default schema name can be inquired.

    >>> dialect.server_version_info >= (1, 0, 0)
    True

> Check if schema exists:

Reviewer 2: ```suggestion
Check if a schema exists:
```

    >>> dialect.has_schema(connection, 'doc')
    True

> Check if table exists:

Reviewer 2: ```suggestion
Check if a table exists:
```
